### PR TITLE
#12: Container now waits until UDF language container is ready.

### LINKED ISSUE #12: java.sql.SQLException: BucketFS: File '/exaudf/exaudfclient' does not exist
## Problem

While using the framework I encountered the problem which is described here:
https://github.com/EXASOL/docker-db/issues/2

How to reproduce:

...
1. statement.execute("CREATE CONNECTION")
2. statement.execute("CREATE OR REPLACE JAVA ADAPTER SCRIPT")
3. statement.execute("CREATE VIRTUAL SCHEMA") -> it comes from this line.
...

It's currently 'fixed' with `sleep`:
...
1. statement.execute("CREATE CONNECTION")
2. statement.execute("CREATE OR REPLACE JAVA ADAPTER SCRIPT")
TimeUnit.SECONDS.sleep(20); 
3. statement.execute("CREATE VIRTUAL SCHEMA") -> it comes from this line.
...

15 seconds of sleep is still not enough, but it wasn't reproduces with 20 seconds.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #12.

### CODE REVIEW

File: src/main/java/com/exasol/containers/wait/LogFileEntryWaitStrategy.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        super();
+        this.container = container;
+        this.logPath = logPath;
+        this.logNamePattern = logNamePattern;
+        this.pattern = pattern;
+    }
+
+    @Override
+    protected void waitUntilReady() {
+        final long expiry = System.currentTimeMillis() + (60 * ONE_SECOND_IN_MILLISECONDS);
```
Comment (morazow) [inline src/main/java/com/exasol/containers/wait/LogFileEntryWaitStrategy.java]: Could you please make this into a constant? I guess, making it `WAIT_DURATION_IN_MILLIS` or similar would be helpful.

File: src/main/java/com/exasol/containers/ExasolContainer.java [hunk 2]
```diff
@@ -7,21 +7,21 @@
...
 
-// [external->dsn~testcontainer-framework-controls-docker-image-download~2]
+// [external->dsn~testcontainer-framework-controls-docker-image-download~1]
 // [impl->dsn~exasol-container-controls-docker-container~1]
 
 @SuppressWarnings("squid:S2160") // Superclass adds state but does not override equals() and hashCode().
 public class ExasolContainer<T extends ExasolContainer<T>> extends JdbcDatabaseContainer<T> {
-    private static final Logger LOGGER = LoggerFactory.getLogger(ExasolContainer.class);
+    private static final String BUCKETFS_DAEMON_LOG_FILENAME_PATTERN = "bucketfsd.*.log";
+    private static final String SCRIPT_LANGUGAGE_CONTAINER_READY_PATTERN = "ScriptLanguages.*extracted$";
```
Comment (morazow) [inline src/main/java/com/exasol/containers/ExasolContainer.java]:
```suggestion
    private static final String SCRIPT_LANGUAGE_CONTAINER_READY_PATTERN = "ScriptLanguages.*extracted$";
```

File: src/main/java/com/exasol/containers/ExasolContainer.java [hunk 3]
```diff
@@ -196,9 +196,25 @@ public Bucket getDefaultBucket() {
...
+    protected void waitUntilContainerStarted() {
+        super.waitUntilContainerStarted();
+        waitUntilUdfLanguageContainerExtracted();
+    }
+
+    // [impl->dsn~exasol-container-ready-criteria~2]
+    private void waitUntilUdfLanguageContainerExtracted() {
+        logger().info("Waiting for UDF language container to be ready.");
+        final LogFileEntryWaitStrategy strategy = new LogFileEntryWaitStrategy(this, EXASOL_CORE_DAEMON_LOGS_PATH,
+                BUCKETFS_DAEMON_LOG_FILENAME_PATTERN, SCRIPT_LANGUGAGE_CONTAINER_READY_PATTERN);
```
Comment (morazow) [inline src/main/java/com/exasol/containers/ExasolContainer.java]:
```suggestion
                BUCKETFS_DAEMON_LOG_FILENAME_PATTERN, SCRIPT_LANGUAGE_CONTAINER_READY_PATTERN);
```

File: src/test/java/com/exasol/containers/ExasolContainerWaitStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                Thread.sleep(2000);
+                writeMessage(this.message);
+            } catch (final InterruptedException exception) {
+                Thread.interrupted();
+            } catch (final IOException exeception) {
+                throw new RuntimeException("Unable to write trigger log message to \"" + this.logFilePath + "\".");
+            }
+        }
+
+        private void writeMessage(final String message2) throws IOException, InterruptedException {
```
Comment (morazow) [inline src/test/java/com/exasol/containers/ExasolContainerWaitStrategyTest.java]: Could you please change `message2` to `messageToWrite` or `messageToAppend`?

File: src/main/java/com/exasol/containers/wait/LogFileEntryWaitStrategy.java (hunk 1 above)
Comment (redcatbear) [reply src/main/java/com/exasol/containers/wait/LogFileEntryWaitStrategy.java]: Done.
Also renamed the polling rate limit.

File: src/test/java/com/exasol/containers/ExasolContainerWaitStrategyTest.java (hunk 4 above)
Comment (redcatbear) [reply src/test/java/com/exasol/containers/ExasolContainerWaitStrategyTest.java]: Done.

**Pull request merged**